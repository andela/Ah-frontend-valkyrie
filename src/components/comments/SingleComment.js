import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Moment from 'react-moment';
import PropTypes from 'prop-types';
import EditComment from './EditComment';
import { deleteComment } from './actions/actions';

export class SingleComment extends Component {
  constructor( props ) {
    super( props );
    this.state = { toggleForm: 'd-none'};
  }

  onClickEditArticle = () => { this.shouldToggleForm(); }

  shouldToggleForm = () => {
    if ( this.state.toggleForm === 'd-block' ) {
      this.setState( { toggleForm: 'd-none' } );
    } else {
      this.setState( { toggleForm: 'd-block' } );
    }
  }

  handleOnDelete = () => {
    const { articleSlug, commentId } = this.props;
    this.props.deleteComment( articleSlug, commentId );
  }

  render() {
    const { commentId, articleSlug, commentBody, username, createdOn, } = this.props;
    const { toggleForm } = this.state;
    return (
      <div>
        <div className="single-comment mt-3">
          <p className="comment-text">{ commentBody }</p>
          <div className="actions">
            <i>
              @
              { username }
            </i>
            &nbsp; &nbsp; &nbsp; (0) &nbsp;
            <i className="fa fa-thumbs-up text-success" />
            &nbsp; &nbsp; &nbsp; (0) &nbsp;
            <i className="fa fa-thumbs-down text-danger" />
            &nbsp; &nbsp; &nbsp;
            { this.props.authUser.isAuthenticated && this.props.authUser.user.username === username ? (
              <Fragment>
                <button type="button" onClick={ this.handleOnDelete } className="btn btn-link text-danger">
                  <i className="fas fa-trash-alt" /> &nbsp;Delete
                </button>
                <button type="button" className="btn btn-link text-primary" onClick={ this.onClickEditArticle } >
                  <i className="fas fa-edit" /> &nbsp;Edit
                </button>
              </Fragment>
            ) : '' }
            &nbsp; &nbsp; &nbsp;
            <Link to="/" className="text-primary" />
            &nbsp; &nbsp;
            <span className="text-muted text-small">
              <i>
              <span>Added:</span> &nbsp;
              <Moment  parse="YYYY-MM-DD HH:mm" > { createdOn } </Moment>
              </i>
            </span>
          </div>
          <EditComment
            commentId={ commentId }
            articleSlug={ articleSlug }
            toggleForm={ toggleForm }
            commentBody={ commentBody }
          />
        </div>
      </div>
    );
  }
}

SingleComment.propTypes = {
  comment: PropTypes.object.isRequired,
  articleSlug: PropTypes.string.isRequired,
  commentId: PropTypes.number.isRequired,
  commentBody: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  createdOn: PropTypes.string.isRequired,
  deleteComment: PropTypes.func.isRequired,
};

const mapStateToProps = state => ( {
  authUser: state.loginReducer,
} );

export default connect( mapStateToProps, { deleteComment } )( SingleComment );