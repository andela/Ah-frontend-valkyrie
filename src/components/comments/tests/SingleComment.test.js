import React from "react";
import { shallow } from "enzyme";
import { SingleComment } from "../SingleComment";

describe("<Single Comment />", () => {
  let mockComponent;
  const props = {
    toggleForm: "d-none",
    authUser: {
      isAuthenticated: true,
      user: {
        username: "testUsername",
      },
    },
    deleteComment: jest.fn(),
    onClickEditArticle: jest.fn(),
  };
  beforeEach(() => {
    mockComponent = shallow(<SingleComment {...props} />);
  });

  it("renders as expected", () => {
    expect(mockComponent).toMatchSnapshot();
  });

  it("onclick handler performs as expected", () => {
    mockComponent.setState({ toggleForm: "d-block" });
    mockComponent.instance().shouldToggleForm();
    expect(mockComponent.state("toggleForm")).toBe("d-none");

    mockComponent.setState({ toggleForm: "d-none" });
    mockComponent.instance().shouldToggleForm();
    expect(mockComponent.state("toggleForm")).toBe("d-block");
  }),

  it("handle submit", () => {
    mockComponent.instance().handleOnDelete();
  });

  it("handle onclick", () => {
    mockComponent.instance().onClickEditArticle();
  });
});
