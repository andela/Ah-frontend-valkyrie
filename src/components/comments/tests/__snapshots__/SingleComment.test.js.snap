// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Single Comment /> renders as expected 1`] = `ShallowWrapper {}`;
